### DISCUSSION TITLE
✅ Set schema name using `.meta({ id })`

### DISCUSSION THREAD

**Discussion by arlyon** [RobinTail/express-zod-api#1393] [Ideas]
Hello

The default schema naming is verbose, and I would like to be able to tune the results. I looked at potentially overriding some method on Documentation however no such luck. An alternative would be to be able to tune the naming with a withMeta call with the added benefit that we can deduplicate identical schemas used across multiple resources using reference equality.

```ts
withMeta(z.object({ data: z.number() })).example("The data").name("MyCoolRequestData");
```

Thanks for this great lib! (and happy holidays, if you celebrate)

Alex

**Comment by RobinTail** [comment]
Hello @arlyon .

I'd rather consider reusing `operationId` property defined within `.build()`, concatenated by something like "RequestSchema" and "ResponseSchema"

https://github.com/RobinTail/express-zod-api/blob/30a69d9b4e3b8fc349eedf03ca84796e51c66972/src/endpoints-factory.ts#L38

**Comment by RobinTail** [comment]
@arlyon ,

I'm going to improve the configuration of the descriptions in the next major release (v16).
See #1416 for details.

Those descriptions do also affect the component naming in the generated documentation (when the `composition` option is set to `components`).

As described above, for each of your endpoint you can specify its `operationId`.
The same `operationId` you can reuse for shaping the component descriptions:

```ts
new Documentation({
  composition: "components", 
  descriptions: {
    requestBody: () => "the body of request",
    requestParameter: ({ method, path }) => `parameter of ${method} ${path}`,
    negativeResponse: ({ operationId }) => `very negative response of ${operationId}`, // <——— this one
    positiveResponse: ({ path }) => `Super positive response of ${path}`,
  },
})
```

So the component name will be the clean id shaped from the description you can make out of its properties and additional alphanumeric assets of your choice.

I believe it provides enough flexibility for your needs.

**Reply by RobinTail** [reply]
You can try it in the dev preview version `16.0.0-beta6`

**Reply by arlyon** [reply]
Nice, thanks!

**Comment by Upsilon-Iridani** [comment]
I too encountered a situation where I wish to set the schema name. I have an interface that is recurring within multiple endpoints, and I figured I could set `mySchema.meta({ id: 'MyId' })`, to get the schema extracted from the response schema as a separate schema with a reference to it. However, it is unclear how the name is determined. Right now the sub schema is called "Schema1", and I can't figure out how to change it.

It makes a huge difference in the readability of both the raw openapi-doc and the generated client code if you keep the schemas tidy :)

Love the library btw!

**Reply by RobinTail** [reply]
Hello @Upsilon-Iridani ,

The extraction of a schema that occurs multiple times during the depiction of a certain request or response is performed by Zod 4:

```ts
const schema = z.string().min(2).meta({ id: "MySchema" });

console.log(
  z.toJSONSchema(
    z.object({
      a: schema,
      b: schema,
    }),
  ),
);
```

```js
{
  '$schema': 'https://json-schema.org/draft/2020-12/schema',
  type: 'object',
  properties: {
    a: { '$ref': '#/$defs/MySchema' },
    b: { '$ref': '#/$defs/MySchema' }
  },
  required: [ 'a', 'b' ],
  additionalProperties: false,
  '$defs': { MySchema: { id: 'MySchema', type: 'string', minLength: 2 } }
}
```

During the generation of the Documetation, having the depictions of multiple Endpoints, those extracted definitions get enumerated by Express Zod API because they may NOT have the unique `id` (circular reference case), OR the `id` you set manually could be equal to the one that was generated.

Though, I'm gonna check out what I could do in order to address this.

**Reply by RobinTail** [reply]
#3133

**Reply by RobinTail** [reply]
🚀 The feature is now available in v26.2.0.
Thank you @arlyon  and @Upsilon-Iridani for your contributions!
